server/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java

```diff
@@ -550,56 +461,62 @@ static ReplicaItemExecutionMode replicaItemExecutionMode(final BulkItemRequest r
-        return executeOnPrimaryWhileHandlingMappingUpdates(primary.shardId(), request.type(),
+        final DeleteRequest request = context.getRequestToExecute();
+        final IndexShard primary = context.getPrimary();
+        executeOnPrimaryWhileHandlingMappingUpdates(context,
             () -> primary.applyDeleteOperationOnPrimary(request.version(), request.type(), request.id(), request.versionType()),
             e -> new Engine.DeleteResult(e, request.version()),
-            mappingUpdater);
+            context::markOperationAsCompleted,
+            mapping -> mappingUpdater.updateMappings(mapping, primary.shardId(), request.type()));
     }
 
-    private static <T extends Engine.Result> T executeOnPrimaryWhileHandlingMappingUpdates(ShardId shardId, String type,
-                                                                                           CheckedSupplier<T, IOException> toExecute,
-                                                                                           Function<Exception, T> onError,
-                                                                                           MappingUpdatePerformer mappingUpdater)
+    private static <T extends Engine.Result> void executeOnPrimaryWhileHandlingMappingUpdates(
+        PrimaryExecutionContext context, CheckedSupplier<T, IOException> toExecute,
+        Function<Exception, T> exceptionToResult, Consumer<T> onComplete, Consumer<Mapping> mappingUpdater)
         throws IOException {
         T result = toExecute.get();
         if (result.getResultType() == Engine.Result.Type.MAPPING_UPDATE_REQUIRED) {
             // try to update the mappings and try again.
             try {
-                mappingUpdater.updateMappings(result.getRequiredMappingUpdate(), shardId, type);
+                mappingUpdater.accept(result.getRequiredMappingUpdate());
             } catch (Exception e) {
                 // failure to update the mapping should translate to a failure of specific requests. Other requests
                 // still need to be executed and replicated.
-                return onError.apply(e);
+                onComplete.accept(exceptionToResult.apply(e));
+                return;
             }
 
+            // TODO - we can fall back to a wait for cluster state update but I'm keeping the logic the same for now
             result = toExecute.get();
 
             if (result.getResultType() == Engine.Result.Type.MAPPING_UPDATE_REQUIRED) {
                 // double mapping update. We assume that the successful mapping update wasn't yet processed on the node
                 // and retry the entire request again.
-                throw new ReplicationOperation.RetryOnPrimaryException(shardId,
-                    "Dynamic mappings are not available on the node that holds the primary yet");
+                context.markAsRequiringMappingUpdate();
```

Comment:
I wonder if we should just mark it as failed here. This would mean that we wouldn't have to introduce the REQUIRED_MAPPING_UPDATE state in this PR and also wouldn't need that retry logic in the executeBulkItem. Note that we should (at the moment, with the MappingUpdatePerformer waiting on ack) only even get to this code here in case where we got an ack, but the node somehow did not apply the cluster state (i.e. super super edge casy).

Reply:
> only even get to this code here in case where we got an ack, but the node somehow did not apply the cluster state

I'm not sure about that - how about when the node was slow to process previous cluster states (where the ack failed) and then submitted another mapping request that turns out to be a noop? It may be that the code current force publishes another cluster state, but I feel that's too brittle to rely on. Further, it may change in the future without us noticing it. Last, I think we will end up introducing the retry loop in a future PR (when we don't wait on ack of all nodes) and I'm doubting it's worth it to take it out now and put it back in (this is the main point of this refactoring).

Reply:
ok, makes sense